src/node.cc

```diff
@@ -1296,86 +1296,160 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
+  bool failed_ = false;
+  bool pushed_ids_ = false;
+  bool closed_ = false;
+
+  // This is only for backwards compatibility, semantically it's always empty.
+  v8::Local<v8::Value> failure_return_value_;
+};
+
+CallbackScope::CallbackScope(Isolate* isolate,
+                             Local<Value> object,
+                             async_context asyncContext)
+  : private_(new InternalCallbackScope(Environment::GetCurrent(isolate),
+                                       object,
+                                       asyncContext)) { }
+
+CallbackScope::~CallbackScope() {
+  delete private_;
+}
+
+InternalCallbackScope::InternalCallbackScope(Environment* env,
+                                             Local<Value> object,
+                                             const async_context& asyncContext)
+  : env_(env),
+    async_context_(asyncContext),
+    object_(object->IsObject() ? object.As<Object>() : Local<Object>()),
+    callback_scope_(env) {
   // If you hit this assertion, you forgot to enter the v8::Context first.
   CHECK_EQ(env->context(), env->isolate()->GetCurrentContext());
 
-  Local<Object> object;
-
-  Environment::AsyncCallbackScope callback_scope(env);
-  bool disposed_domain = false;
-
-  if (recv->IsObject()) {
-    object = recv.As<Object>();
+  if (env->using_domains()) {
+    CHECK(object->IsObject());
+    failed_ = DomainEnter(env, object_);
+    if (failed_) {
+      failure_return_value_ = Undefined(env->isolate());
+      return;
+    }
   }
 
-  if (env->using_domains()) {
-    CHECK(recv->IsObject());
-    disposed_domain = DomainEnter(env, object);
-    if (disposed_domain) return Undefined(env->isolate());
+  if (asyncContext.async_id != 0) {
+    // No need to check a return value because the application will exit if
+    // an exception occurs.
+    AsyncWrap::EmitBefore(env, asyncContext.async_id);
   }
 
-  MaybeLocal<Value> ret;
+  env->async_hooks()->push_ids(async_context_.async_id,
+                               async_context_.trigger_async_id);
+  pushed_ids_ = true;
+}
 
-  {
-    AsyncHooks::ExecScope exec_scope(env, asyncContext.async_id,
-                                     asyncContext.trigger_async_id);
+InternalCallbackScope::~InternalCallbackScope() {
+  Close();
+}
 
-    if (asyncContext.async_id != 0) {
-      // No need to check a return value because the application will exit if
-      // an exception occurs.
-      AsyncWrap::EmitBefore(env, asyncContext.async_id);
-    }
+void InternalCallbackScope::Close() {
+  if (closed_) return;
+  closed_ = true;
 
-    ret = callback->Call(env->context(), recv, argc, argv);
+  if (pushed_ids_)
+    env_->async_hooks()->pop_ids(async_context_.async_id);
 
-    if (ret.IsEmpty()) {
-      // NOTE: For backwards compatibility with public API we return Undefined()
-      // if the top level call threw.
-      return callback_scope.in_makecallback() ?
-          ret : Undefined(env->isolate());
-    }
+  if (failed_) return;
```

Comment:
This matches the previous behaviour, and I think that’s okay because you already explicitly accounted for that situation in `pop_ids()`.